src/app/modules/form/public-form/public-form.middlewares.ts

```diff
@@ -3,7 +3,32 @@ import { StatusCodes } from 'http-status-codes'
 
 import { WithForm } from '../../../../types'
 import { FormDeletedError } from '../form.errors'
-import { isFormPublic } from '../form.service'
+import {
+  checkFormSubmissionLimitAndDeactivateForm,
+  isFormPublic,
+} from '../form.service'
+
+/**
+ * Express middleware function that checks if a form has exceeded its submission limits before allowing
+ * downstream middleware to handle the request. Otherwise, it returns a HTTP 404.
+ */
+export const checkFormSubmissionLimitAndDeactivate: RequestHandler = async (
+  req,
+  res,
+  next,
+) => {
+  const { form } = req as WithForm<typeof req>
+  const formResult = await checkFormSubmissionLimitAndDeactivateForm(form)
+  if (!formResult.isErr()) {
+    return next()
+  } else {
+    return res.status(StatusCodes.NOT_FOUND).json({
+      message: form.inactiveMessage,
+      isPageFound: true, // Flag to prevent default 404 subtext ("please check link") from showing
+      formTitle: form.title,
+    })
+  }
```

Comment:
pardon my nitpick, i thought this would be slightly neater
```suggestion
  if (formResult.isErr()) {
    return res.status(StatusCodes.NOT_FOUND).json({
      message: form.inactiveMessage,
      isPageFound: true, // Flag to prevent default 404 subtext ("please check link") from showing
      formTitle: form.title,
    })
  }
  
  return next()
```

Reply:
also I think we should use HTTP 429 Too Many Requests here instead of 404 Not Found

Reply:
So I think the point here was that the form should be indistinguishable from being manually deactivated? Even if we return a 429 here, any follow-up request will get a 404 since the form public check comes before this under all routes.